### PR TITLE
feat(ibis): Implement Snowflake Metadata APIs

### PR DESCRIPTION
**Pull request by ongdisheng**
Implement Snowflake Metadata APIs [#709](https://github.com/Canner/WrenAI/issues/709)

### Request payload example
```json
{
  "connectionInfo": {
    "user": "my-user",
    "password": "my-pwd",
    "account": "my-account",
    "database": "my-db",
    "schema": "my-schema"
  }
}
```

### Response example

- name: unique table name
- columns
     - name: column name in the datasource
     - type: column data type
     - notNull: boolean, nullable or not
     - description: column description(comment) if any
     - properties: column properties if any
- description: table description if any 
- properties
     - schema: schema name to build tableReference
     - catalog: catalog name to build tableReference
     - table: table name to build tableReference
- primaryKey: the column name which is bind with primary constraint    

```json
[
    {
        "name": "db.REGION",
        "columns": [
            {
                "name": "R_COMMENT",
                "type": "VARCHAR",
                "notNull": false,
                "description": "",
                "properties": null
            }
        ],
        "description": "",
        "properties": {
            "schema": "db",
            "catalog": "def",
            "table": "REGION"
        },
        "primaryKey": ""
    },
]
```

### Request payload example
```json
{
  "connectionInfo": {
    "user": "my-user",
    "password": "my-pwd",
    "account": "my-account",
    "database": "my-db",
    "schema": "my-schema"
  }
}
```

### Response example
- constraints
     - constraintName: unique constraint name
     ```({constraint_table}_{constraint_column}_{constrainted_table}_{constrainted_column})```
     - constraintType: “FOREIGN KEY”
     - constraintTable
     - constraintColumn
     - constraintedTable
     - constraintedColumn
```json
[
    {
        "constraintName": "ORDERS_O_CUSTKEY_CUSTOMER_C_CUSTKEY",
        "constraintType": "FOREIGN KEY",
        "constraintTable": "db.ORDERS",
        "constraintColumn": "O_CUSTKEY",
        "constraintedTable": "db.CUSTOMER",
        "constraintedColumn": "C_CUSTKEY"
    },
]
```

### CODE REVIEW

**Comment by ongdisheng** [comment]
Both Ruff lint and pytest checks have passed successfully
Lint

Pytest

**Comment by ongdisheng** [comment]
Hi @grieve54706 @goldmedal , could you take a look at my PR when you have some free time? I'd really appreciate your feedback, thanks in advance!

**Review by goldmedal — APPROVED** [review]
Thanks @ongdisheng. Overall looks good to me 👍 . Let's wait for @grieve54706 to review it.

File: ibis-server/tests/routers/v2/connector/test_snowflake.py [hunk 1]
```diff
@@ -261,14 +261,24 @@ def test_validate_rule_column_is_valid_without_one_parameter(manifest_str):
...
         assert response.text == "Missing required parameter: `modelName`"
 
-    @pytest.mark.skip(reason="Not implemented")
     def test_metadata_list_tables():
-        pass
+        response = client.post(
+            url=f"{base_url}/metadata/tables",
+            json={"connectionInfo": connection_info},
+        )
+        assert response.status_code == 200
```
Comment (grieve54706) [inline ibis-server/tests/routers/v2/connector/test_snowflake.py:269]: Please add more expected to like `test_trino`.

File: ibis-server/tests/routers/v3/connector/snowflake/test_functions.py [hunk 2]
```diff
@@ -58,14 +58,14 @@ def test_function_list():
         assert response.status_code == 200
         result = response.json()
         assert len(result) == 427
-        the_func = next(filter(lambda x: x["name"] == "abs", result))
+        the_func = next(filter(lambda x: x["name"] == "ABS", result))
```
Comment (grieve54706) [inline ibis-server/tests/routers/v3/connector/snowflake/test_functions.py:61]: We make the function name lowercase. You can follow up on it.

**Review by grieve54706 — CHANGES_REQUESTED** [review]
Well done. Thanks for your contribution. Welcome.

**Comment by ongdisheng** [comment]
@goldmedal @grieve54706, thanks for the quick feedback! I’ll have the fixes ready later today.

File: ibis-server/tests/routers/v2/connector/test_snowflake.py (hunk 1 above)
Comment (ongdisheng) [reply ibis-server/tests/routers/v2/connector/test_snowflake.py:269]: I’ve updated the test. Could you please take a look when you have time and let me know if I’ve done it correctly? Thanks @grieve54706!

File: ibis-server/tests/routers/v2/connector/test_snowflake.py [hunk 3]
```diff
@@ -261,14 +261,45 @@ def test_validate_rule_column_is_valid_without_one_parameter(manifest_str):
...
+        }
 
-    @pytest.mark.skip(reason="Not implemented")
     def test_metadata_list_constraints():
-        pass
+        response = client.post(
+            url=f"{base_url}/metadata/constraints",
+            json={"connectionInfo": connection_info},
+        )
+        assert response.status_code == 200
```
Comment (grieve54706) [inline ibis-server/tests/routers/v2/connector/test_snowflake.py:297]: Hi @ongdisheng,
Could you add expected like `test_metadata_list_tables`?
You did a good job. Thanks for your help.

Comment (ongdisheng) [reply ibis-server/tests/routers/v2/connector/test_snowflake.py:297]: Sure, no problem.

Comment (ongdisheng) [reply ibis-server/tests/routers/v2/connector/test_snowflake.py:297]: Hi @grieve54706, I've noticed that the ```TPCH_SF1``` schema doesn’t seem to enforce foreign key constraints, which causes the ```SHOW IMPORTED KEYS``` command to return an empty list.

Here are the results I found when querying the ```TABLE_CONSTRAINTS```:
image

Given this, would it be appropriate to add an assertion to check that the result returns an empty list? I’d be glad to hear any other suggestions you may have. Thank you!

Comment (grieve54706) [reply ibis-server/tests/routers/v2/connector/test_snowflake.py:297]: Yes, asserting it is an empty list is a good choice.

**Review by grieve54706 — APPROVED** [review]
LGTM. Thanks @ongdisheng.

**Comment by goldmedal** [comment]
Thanks @ongdisheng and @grieve54706 for reviewing 👍

**Pull request merged**